Menu action to export a watchonly wallet

achow101 opened the pull request:
Allows a user to export a watchonly version of their wallet to be used in an airgapped setup.

Built on https://github.com/bitcoin/bitcoin/pull/32489

achow101 committed:
wallet: Move listdescriptors retrieving from RPC to CWallet

When listdescriptors retrieves the descriptors from the wallet, instead
of having this logic in the RPC, move it into CWallet itself. This
will enable other functions to get the descriptors in an exportable
form.

achow101 committed:
wallet: Add CWallet::ExportWatchOnly

ExportWatchOnly produces a watchonly wallet file from a CWallet. This
can be restored onto another instance of Bitcoin Core to allow that
instance to watch the same descriptors, and also have all of the same
initial address book and transactions.

achow101 committed:
wallet, rpc: Add exportwatchonlywallet RPC

achow101 committed:
descriptor: Add CanSelfExpand()

CanSelfExpand() reports whether a descriptor can be expanded without
needing any caches or private keys to be provided by the caller of
Expand().

achow101 committed:
wallet: Use Descriptor::CanSelfExpand() in CanGetAddresses()

If a descriptor does not need any caches or private keys in order to
expand, then CanGetAddresses() should return true for that descriptor.

achow101 committed:
test: Test for exportwatchonlywallet

achow101 committed:
wallet: Write new descriptor's cache in AddWalletDescriptor

If a new WalletDescriptor is provided to us with a cache, write the
cache to disk as well.

achow101 committed:
gui: Menu action for exporting a watchonly wallet

achow101 committed:
gui: Menu action for exporting a watchonly wallet

polespinasa committed:
doc: update offline-signing-tutorial to use exportwatchonlywallet rpc

pablomartin4btc reviewed (commented):
The base PR (bitcoin/bitcoin#32489) has been merged, this will need a rebase.

Reviewed the GUI commit (59fd6f9):
- The action is enabled for all wallets. Please consider disabling it for already-watchonly wallets — here cdd15e1775ba316f1a2e3a44921fc7c6fbabd137 you have a Qt test covering the action verifying it's enabled/disabled correctly per wallet type.
- Missing release notes for a new user-facing menu action.
- Please consider the [comment](https://github.com/bitcoin/bitcoin/pull/32489#discussion_r3518302122) from @Sjors regarding `CHECK_NONFATAL` (if it's still relevant).

achow101 committed:
doc: Release note for export watchonly wallet gui action

achow101 committed:
gui: Add ExceptionSafeConnect that takes a lambda

There is a variant of QObject::connect which takes 3 arguments. Add a
variant of ExceptioNSafeConnect that does the same thing.

polespinasa commented on src/qt/bitcoingui.cpp:
in a4f67c85c6835724009cb6b6336c93288c324336 gui: Menu action for exporting a watchonly wallet

If you close all wallets, the option to export a watch-only-wallet remains usable, but it should be grey and unusable.
Fix:

```diff
$ git diff
diff --git a/src/qt/bitcoingui.cpp b/src/qt/bitcoingui.cpp
index 690ad86390..ad1ecf572d 100644
--- a/src/qt/bitcoingui.cpp
+++ b/src/qt/bitcoingui.cpp
@@ -882,6 +882,7 @@ void BitcoinGUI::setWalletActionsEnabled(bool enabled)
     openAction->setEnabled(enabled);
     m_close_wallet_action->setEnabled(enabled);
     m_close_all_wallets_action->setEnabled(enabled);
+    m_export_watchonly_action->setEnabled(enabled);
 }
 
 void BitcoinGUI::createTrayIcon()

```

achow101 replied:
Done

polespinasa commented on src/wallet/interfaces.cpp:
```diff
@@ -522,6 +523,11 @@ class WalletImpl : public Wallet
     CWallet* wallet() override { return m_wallet.get(); }
 
+    util::Result<std::string> exportWatchOnlyWallet(const fs::path& destination) override {
+        LOCK(m_wallet->cs_wallet);
```
in a4f67c8 gui: Menu action for exporting a watchonly wallet

Shouldn't here call `TopUpKeyPool()` like we do in the RPC?¿

achow101 replied:
Done

polespinasa commented on src/qt/bitcoingui.cpp:
```diff
@@ -524,6 +528,16 @@ void BitcoinGUI::createActions()
+        GUIUtil::ExceptionSafeConnect(m_export_watchonly_action, &QAction::triggered, [this](bool) {
+            QString destination = GUIUtil::getSaveFileName(this, tr("Save Watch-only Wallet Export"), QString(), QString(), nullptr);
+            if (destination.isEmpty()) return;
+            util::Result<std::string> export_res = walletFrame->currentWalletModel()->wallet().exportWatchOnlyWallet(GUIUtil::QStringToPath(destination));
```
in a4f67c8 gui: Menu action for exporting a watchonly wallet

What if `walletFrame->currentWalletModel()` is `nullptr`?
That causes a segmentation fault:

<img width="766" height="48" alt="Image" src="https://github.com/user-attachments/assets/46e05391-c5f8-4fcf-bb04-cfdeaee0ca8c" />

This can be triggered when no wallet is loaded and you try to export a watchonly wallet.
Which should not happen but can happen because of https://github.com/bitcoin-core/gui/pull/872/changes/a4f67c85c6835724009cb6b6336c93288c324336#r3664616220

We could probably also protect this here with:

```diff
$ git diff
diff --git a/src/qt/bitcoingui.cpp b/src/qt/bitcoingui.cpp
index 690ad86390..eee44330dd 100644
--- a/src/qt/bitcoingui.cpp
+++ b/src/qt/bitcoingui.cpp
@@ -531,7 +531,9 @@ void BitcoinGUI::createActions()
         GUIUtil::ExceptionSafeConnect(m_export_watchonly_action, &QAction::triggered, [this](bool) {
             QString destination = GUIUtil::getSaveFileName(this, tr("Save Watch-only Wallet Export"), QString(), QString(), nullptr);
             if (destination.isEmpty()) return;
-            util::Result<std::string> export_res = walletFrame->currentWalletModel()->wallet().exportWatchOnlyWallet(GUIUtil::QStringToPath(destination));
+            WalletModel* model = walletFrame->currentWalletModel();
+            if (!model) return;
+            util::Result<std::string> export_res = model->wallet().exportWatchOnlyWallet(GUIUtil::QStringToPath(destination));
             if (export_res) {
                 QMessageBox::information(nullptr, tr("Export Successful"), tr("The wallet has been exported to ") + QString::fromStdString(*export_res));
             } else {

```

achow101 replied:
The model should never be nullptr. Added an `Assume`

pablomartin4btc reviewed (commented):
ACK 6573196e63bf4113d332c802f409e3336aebab41

hebasto reviewed (approved):
ACK 6573196e63bf4113d332c802f409e3336aebab41.
